packages/flutter/lib/src/material/pickers/calendar_date_range_picker.dart

```diff
@@ -0,0 +1,753 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:math' as math;
+
+import 'package:flutter/gestures.dart' show DragStartBehavior;
+import 'package:flutter/rendering.dart';
+import 'package:flutter/services.dart';
+import 'package:flutter/widgets.dart';
+
+import '../color_scheme.dart';
+import '../divider.dart';
+import '../material_localizations.dart';
+import '../text_theme.dart';
+import '../theme.dart';
+
+import 'date_utils.dart' as utils;
+
+const double _monthItemHeaderHeight = 58.0;
+const double _monthItemFooterHeight = 12.0;
+const double _monthItemRowHeight = 42.0;
+const double _monthItemSpaceBetweenRows = 8.0;
+const double _horizontalPadding = 8.0;
+const double _maxCalendarWidthLandscape = 384.0;
+const double _maxCalendarWidthPortrait = 480.0;
+
+/// Displays a scrollable calendar grid that allows a user to select a range
+/// of dates.
+///
+/// Note: this is not publicly exported (see pickers.dart), as it is an
+/// internal component used by [showDateRangePicker].
+class CalendarDateRangePicker extends StatefulWidget {
+  /// Creates a scrollable calendar grid for picking date ranges.
+  CalendarDateRangePicker({
+    Key key,
+    this.initialStartDate,
+    this.initialEndDate,
+    @required this.firstDate,
+    @required this.lastDate,
+    DateTime currentDate,
+    @required this.onStartDateChanged,
+    @required this.onEndDateChanged,
+  }) : currentDate = utils.dateOnly(currentDate ?? DateTime.now()),
```

Comment:
All of the other parameters can be null too? 

If they're not null, assert that lastDate >= firstDate etc?